unit_tests/kernels/UnitTestKernelUtils.C

```diff
@@ -382,6 +385,417 @@ void calc_mass_flow_rate_scs(
+
+  auto meBC = sierra::nalu::get_surface_master_element(topo);
+
+  dataNeeded.add_cvfem_surface_me(meBC);
+  dataNeeded.add_coordinates_field(coordinates, ndim, sierra::nalu::CURRENT_COORDINATES);
+  dataNeeded.add_gathered_nodal_field(scalarField, 1);
+  dataNeeded.add_gathered_nodal_field(dnv, 1);
+  dataNeeded.add_gathered_nodal_field(gradField, ndim);
+  dataNeeded.add_master_element_call(sierra::nalu::SCS_AREAV, sierra::nalu::CURRENT_COORDINATES);
+
+  const stk::mesh::Selector selector = meta.locally_owned_part() | meta.globally_shared_part();
+  const auto& buckets = bulk.get_buckets(meta.side_rank(), selector);
+
+  const int bytes_per_team = 0;
+  const int bytes_per_thread = sierra::nalu::get_num_bytes_pre_req_data(dataNeeded, meta.spatial_dimension()) ;
+
+  auto v_shape_function = Kokkos::View<double**>("shape_function", meBC->numIntPoints_, meBC->nodesPerElement_);
+
+  auto team_exec = sierra::nalu::get_team_policy(buckets.size(), bytes_per_team, bytes_per_thread);
+
+  Kokkos::parallel_for(team_exec, [&](const sierra::nalu::TeamHandleType& team) {
+    auto& b = *buckets[team.league_rank()];
+    const auto length = b.size();
+
+    EXPECT_EQ(b.topology(), topo);
+
+    sierra::nalu::ScratchViews preReqData(
+      team, bulk, topo, dataNeeded);
+
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, length), [&](const size_t& k) {
+      stk::mesh::Entity face = b[k];
+      sierra::nalu::fill_pre_req_data(dataNeeded, bulk, topo, face, preReqData);
+
+      meBC->shape_fcn(v_shape_function.data());
+      auto v_dnv = preReqData.get_scratch_view_1D(dnv);
+      auto v_scalar = preReqData.get_scratch_view_1D(scalarField);
+      auto v_scs_areav = preReqData.get_me_views(sierra::nalu::CURRENT_COORDINATES).scs_areav;
+      const stk::mesh::Entity* node_rels = preReqData.elemNodes;
+      const int* ipNodeMap = meBC->ipNodeMap();
+
+      for (int ip = 0; ip < meBC->numIntPoints_; ++ip) {
+        double qIp = 0.0;
+        for (int n = 0; n < meBC->nodesPerElement_; ++n) {
+          qIp += v_shape_function(ip, n) * v_scalar(n);
+        }
+
+        const int nn = ipNodeMap[ip];
+        double* dqdxNN = stk::mesh::field_data(gradField, node_rels[nn]);
+
+        for (int d = 0; d < ndim; ++d) {
+          double fac  = qIp * v_scs_areav(ip, d) / v_dnv(nn);
+          Kokkos::atomic_add(dqdxNN + d, fac);
+        }
+      }
+    });
+  });
+}
+
+void calc_projected_nodal_gradient_boundary(
+  const stk::mesh::BulkData& bulk,
+  const stk::topology& topo,
+  const VectorFieldType& coordinates,
+  const ScalarFieldType& dnv,
+  const VectorFieldType& vectorField,
+  const GenericFieldType& gradField)
+{
+  const auto& meta = bulk.mesh_meta_data();
+  const int ndim = meta.spatial_dimension();
+  EXPECT_EQ(ndim, 3);
+  EXPECT_EQ(topo.rank(), meta.side_rank());
+
+  sierra::nalu::ElemDataRequests dataNeeded;
+
+  auto meBC = sierra::nalu::get_surface_master_element(topo);
+
+  dataNeeded.add_cvfem_surface_me(meBC);
```

Comment:
I find this interesting... If you had a face:element algorithm, would things work out of the box? For example, if we also needed to gather interior coordinates could we do that? Probably not, right?